eclipse-scout-core/src/table/Table.ts

```diff
@@ -5713,15 +5705,172 @@ export class Table extends Widget implements TableModel, Filterable<TableRow> {
+    if (arrays.equals(this.columns, columnOrModels)) {
+      return;
+    }
+    this._setColumns(columnOrModels, initValue);
+    if (this.rendered) {
+      this._renderColumns();
+    }
+  }
+
+  protected _setColumns(columnOrModels: ObjectOrChildModel<Column<any>>[], initValue?: any | ((Column, TableRow) => any)) {
+    // ensure initValue function
+    if (initValue === undefined) {
+      initValue = null;
+    }
+    if (!objects.isFunction(initValue)) {
+      const value = initValue;
+      initValue = (column: Column, row: TableRow) => value;
+    }
+
+    // ensure columns
+    const columns = columnOrModels
+      .map(column => this._ensureColumn(column))
+      .filter(column => !!column);
+
+    const columnsOld = [...this.columns];
+    const columnsDeleted = this.columns.filter(column => column instanceof Column && !arrays.contains(columns, column));
+    const columnsInserted = columns.filter(column => !arrays.contains(this.columns, column));
+    const columnsCurrent = [...this.columns];
+
+    let buffering: JQuery.Deferred<void>;
+
+    // deleted columns
+    for (const column of columnsDeleted) {
+      // update index of other columns and handle cells if removed column has an index set
+      if (column.index >= 0) {
+        // update index of all columns whose index is greater than the index of the removed column
+        for (const c of columnsCurrent) {
+          if (c.index > column.index) {
+            c.index--;
+          }
+        }
+
+        // remove cells for all rows at the correct position (column.index)
+        this.rows.forEach(row => row.cells.splice(column.index, 1));
+
+      }
+
+      // destroy column and remove from current columns
+      column.destroy();
+      arrays.remove(columnsCurrent, column);
+    }
+
+    // inserted columns
+    for (const column of columnsInserted) {
+      // set index to max index + 1 if no index is provided
+      if (column.index < 0) {
+        column.index = Math.max(-1, ...columnsCurrent.map(column => column.index)) + 1;
+      }
+
+      // update index of all columns whose index is greater or equal to the index of the new column
+      for (const c of columnsCurrent) {
+        if (c.index >= column.index) {
+          c.index++;
+        }
+      }
+
+      // set updateBuffer buffering until all columns are added, otherwise column.initCell triggers updateRow before the rows cell array is modified
+      if (!buffering) {
+        buffering = $.Deferred();
+        this.updateBuffer.pushPromise(buffering.promise());
+      }
+
+      // insert new cell for all rows at the correct position (column.index)
+      this.rows.forEach(row => {
+        arrays.insert(row.cells, column.initCell(initValue(column, row), row), column.index);
+      });
+
+      // add to current columns
+      columnsCurrent.push(column);
+    }
+
+    this._setProperty('columns', columns);
```

Comment:
The logic of _initColumns needs to applied here as well.
- gui only columns have to be at the beginning, and it  must not be possible to remove them using setColumns, otherwise the state of checkable, rowIconVisible etc. will be inconsistent 
- permanent sort columns have to be updated
- node icon column needs to be updated
- checkable column needs to be updated
-> extract relevant code from initColumns to a separate method so it can be called here as well

Reply:
Also, there are other column properties, that have an effect on the rows and need to be re-applied, if the new columns have one of these properties set. Sorting, grouping, filters, background-effect, maybe more. Have a look at the Table._init-function, all of these properties are applied there, directly or indirectly. Maybe it is enough to call _sort() (which also groups), _calculateValuesForBackgroundEffect() and _updateRowStructure() with applyFilters only.